sapicore/pipeline/__init__.py

```diff
@@ -1,21 +1,44 @@
-"""Pipeline
-===========
+"""Configurable workflows and scripts.
+
+Pipelines are flexible objects consisting of two logical parts: configurable parameters and a custom
+script defined by overriding the abstract method :meth:`~Pipeline.run`.
+
+To facilitate the development and testing of models, Sapicore provides a compact default simulation
+pipeline with a sample configuration YAML. Advanced users may use those as a basis for custom workflows.
 
-A pipeline is a class that can run the training and testing of a model.
-It should handle command line input, model configuration, and training and
-testing.
 """
+import os
+from tree_config import Configurable, load_config, apply_config
+
+__all__ = ("Pipeline",)
 
-__all__ = ('PipelineBase', )
 
+class Pipeline(Configurable):
+    """Pipeline base class.
+
+    Parameters
+    ----------
+    configuration: dict or str, optional
+        Dictionary or path to YAML to use with :meth:`apply_config`.
 
-class PipelineBase:
-    """Baseclass for pipelines.
     """
 
-    def run(self) -> None:
-        """Runs the pipeline.
+    def __init__(self, configuration: dict | str = None, **kwargs):
+        # parse configuration from YAML if given and use resulting dictionary to initialize attributes.
+        if isinstance(configuration, str) and os.path.exists(configuration):
+            apply_config(self, load_config(None, configuration))
+            self.configuration = load_config(self, filename=configuration)
 
-        To be overwritten by users.
-        """
+        elif isinstance(configuration, dict):
+            apply_config(self, configuration)
+
+        # developer may define arbitrary attributes at instantiation.
+        for key, value in kwargs.items():
```

Comment:
It's not a good idea to arbitrarily eat all kwargs and set them as object props. This will break multi-inheritance because 1. it doesn't use super and 2. subclasses/siblingclasses won't be able to pass kwargs to each other.

It's better to pass kwargs to super. And if someone wants to set object props they'll inherit and manually list the props in their init.

Reply:
I notice you do this in multiple place. I think in general we shouldn't use this pattern of setting all kwargs to class variables but either filter them by config pass it on to super.

Reply:
Fixed for all constructors.